src/wp-includes/block-patterns.php

```diff
@@ -350,6 +350,98 @@ function _register_remote_theme_patterns() {
+	 * let it override any patterns from the parent theme that shares the same slug.
+	 */
+	$themes   = array();
+	$theme    = wp_get_theme();
+	$themes[] = $theme;
+	if ( $theme->parent() ) {
+		$themes[] = $theme->parent();
+	}
+	$registry = WP_Block_Patterns_Registry::get_instance();
+
+	foreach ( $themes as $theme ) {
+		$pattern_data = _wp_get_block_patterns( $theme );
+		$dirpath      = $theme->get_stylesheet_directory() . '/patterns/';
+
+		foreach ( $pattern_data['patterns'] as $file => $pattern_data ) {
+			if ( $registry->is_registered( $pattern_data['slug'] ) ) {
+				continue;
+			}
+
+			// The actual pattern content is the output of the file.
+			ob_start();
+			include $dirpath . $file;
+			$pattern_data['content'] = ob_get_clean();
+			if ( ! $pattern_data['content'] ) {
+				continue;
+			}
+
+			// Translate the pattern metadata.
+			$text_domain = $theme->get( 'TextDomain' );
+			// phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralDomain,WordPress.WP.I18n.LowLevelTranslationFunction
+			$pattern_data['title'] = translate_with_gettext_context( $pattern_data['title'], 'Pattern title', $text_domain );
+			if ( ! empty( $pattern_data['description'] ) ) {
+				// phpcs:ignore WordPress.WP.I18n.NonSingularStringLiteralText,WordPress.WP.I18n.NonSingularStringLiteralDomain,WordPress.WP.I18n.LowLevelTranslationFunction
+				$pattern_data['description'] = translate_with_gettext_context( $pattern_data['description'], 'Pattern description', $text_domain );
+			}
+
+			register_block_pattern( $pattern_data['slug'], $pattern_data );
+		}
+	}
+}
+add_action( 'init', '_register_theme_block_patterns' );
+
+/**
+ * Gets block pattern data for a theme.
+ *
+ * @since 6.4.0
+ * @access private
+ *
+ * @param WP_Theme $theme Theme.
+ * @return array Block pattern data.
+ */
+
+function _wp_get_block_patterns( WP_Theme $theme ) {
+	$option_name    = 'wp_theme_patterns_' . $theme->get_stylesheet();
+	$version        = $theme->get( 'Version' );
+	$can_use_cached = ! wp_is_development_mode( 'theme' );
+
+	if ( $can_use_cached ) {
+		$pattern_data = get_site_option( $option_name );
+		if ( is_array( $pattern_data ) && $pattern_data['version'] === $version ) {
+			return $pattern_data;
+		}
+	}
+
+	$pattern_data = array(
+		'version'  => $version,
+		'patterns' => array(),
+	);
+	$dirpath      = $theme->get_stylesheet_directory() . '/patterns/';
+
+	if ( ! file_exists( $dirpath ) ) {
+		if ( $can_use_cached ) {
+			update_site_option( $option_name, $pattern_data );
+			if ( ! is_multisite() ) {
+				wp_set_option_autoload( $option_name, 'yes' );
+			}
+		}
+		return $pattern_data;
+	}
+	$files = glob( $dirpath . '*.php' );
+	if ( ! $files ) {
+		if ( $can_use_cached ) {
+			update_site_option( $option_name, $pattern_data );
+			if ( ! is_multisite() ) {
+				wp_set_option_autoload( $option_name, 'yes' );
+			}
+		}
+		return $pattern_data;
+	}
```

Comment:
This block is almost the same twice, and it's replicated a third time at the end of this function. It would be simpler to keep the logic to set the result in the cache or database in a single place.

Reply:
Simplier but not better for performance. Exiting early is by design.

Reply:
Neither is better for performance. Exiting early makes sense generally, but not when the only way to do it is duplicating code IMO.